### PR TITLE
Fix dwc access control

### PR DESCRIPTION
**Pull request by AlanWierzchonCA**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [Yes]
| New feature/enhancement? (use the a.x branch)      | [No]
| Deprecations?                          | [No]
| BC breaks? (use the c.x branch)        | [No]
| Automated tests included?              | [Yes] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:

User without privileges can create and delete ’Dynamic Content’
image-2022-04-22-09-52-32-189

#### Steps to test this PR:
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))

* Create new Role with only Dynamic Content Permissions: 
Dynamic Content - User has access to _View Own_ and _View Others_

* Create new User with created Role
* Login with created User
* request GET https://8080-alanwierzchonca-mautic-h31qcblosf3.ws-eu47.gitpod.io/s/dwc/new
* request POST https://8080-alanwierzchonca-mautic-h31qcblosf3.ws-eu47.gitpod.io/s/dwc/delete/1
 

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11278"></a>

### CODE REVIEW

**Review by RCheesley — DISMISSED** [review]
Thanks for the PR @AlanWierzchonCA - please could you reach out to me on Slack at https://mautic.org/slack as I need to chat with you about this issue.

Works as expected, after applying the PR I am correctly getting a 403 if I try to hit the create new / delete.

Please also see the PHPSTAN issues which need some attention :)

File: app/bundles/DynamicContentBundle/Controller/DynamicContentController.php [hunk 1]
```diff
@@ -479,12 +479,17 @@ public function deleteAction($objectId)
             $model  = $this->getModel('dynamicContent');
             $entity = $model->getEntity($objectId);
 
-            if (null === $entity) {
+            if (!$this->get('mautic.security')->isGranted('dynamiccontent:dynamiccontents:deleteown')
+                && !$this->get('mautic.security')->isGranted('dynamiccontent:dynamiccontents:deleteother')) {
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/Controller/DynamicContentController.php]: I'm wondering how the code can know whether the entity can be deleted based on "own" or "other" permissions when the security service doesn't know who created the entity. Shouldn't we use the `hasEntityAccess` method instead?

Isn't that exact check slightly bellow this one on line 493?

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
+<?php
+
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php:2]: New files should declare strict types.
```suggestion

declare(strict_types=1);

```

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        self::PERMISSION_CREATE       => 52,
+        self::PERMISSION_DELETE_OWN   => 66,
+        self::PERMISSION_DELETE_OTHER => 150,
+    ];
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+    }
+
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]: This is unnecessary
```suggestion
```

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    public function testAccessControlNewAction()
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]:
```suggestion
    public function testAccessControlNewAction(): void
```

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    {
+        $this->createAndLoginUser(self::PERMISSION_CREATE);
+        $this->client->request(Request::METHOD_GET, '/s/dwc/new');
+
+        Assert::assertSame(Response::HTTP_OK, $this->client->getResponse()->getStatusCode());
+    }
+
+    public function testForbiddenNewAction()
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]:
```suggestion
    public function testForbiddenNewAction(): void
```

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    {
+        $this->createAndLoginUser();
+        $this->client->request(Request::METHOD_GET, '/s/dwc/new');
+
+        Assert::assertSame(Response::HTTP_FORBIDDEN, $this->client->getResponse()->getStatusCode());
+    }
+
+    public function testAccessDeleteAction()
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]:
```suggestion
    public function testAccessDeleteAction(): void
```

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    {
+        $this->createAndLoginUser(self::PERMISSION_DELETE_OWN);
+        $this->client->request(Request::METHOD_POST, '/s/dwc/delete');
+
+        Assert::assertSame(Response::HTTP_OK, $this->client->getResponse()->getStatusCode());
+    }
+
+    public function testForbiddenDeleteAction()
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]:
```suggestion
    public function testForbiddenDeleteAction(): void
```

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    {
+        $this->createAndLoginUser();
+        $this->client->request('GET', '/s/dwc/delete');
+
+        Assert::assertSame(Response::HTTP_FORBIDDEN, $this->client->getResponse()->getStatusCode());
+    }
+
+    private function createAndLoginUser($permission = null): User
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]:
```suggestion
    private function createAndLoginUser(string $permission = null): User
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Thanks for this! Please check my comments if it makes sense. I also suggested the fixes for PHPSTAN issues.

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php (hunk 2 above)
Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php:2]: OK, I just added the strinct types

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php (hunk 3 above)
Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]: fixed

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php (hunk 5 above)
Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]: fixed

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php (hunk 6 above)
Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]: fixed

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php (hunk 7 above)
Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]: fixed

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php (hunk 8 above)
Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]: fixed

File: app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php (hunk 4 above)
Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Tests/Controller/DynamicContentControllerFunctionalTest.php]: Fixed

File: app/bundles/DynamicContentBundle/Controller/DynamicContentController.php (hunk 1 above)
Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentController.php]: Hi @escopecz, 
on lines 482, 483 the code checks if the logged in user has general permission to delete their own or other dynamic content.
If not, the user should not know if the entity exists or not. It should get 403, not 401.

In line 493 there is a checks base on existting entity Id.

**Comment by AlanWierzchonCA** [comment]
Hi @escopecz, @RCheesley 
thanks for review and tests. 

I just made fixes and answered questions.

File: app/bundles/DynamicContentBundle/Controller/DynamicContentController.php (hunk 1 above)
Comment (escopecz) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentController.php]: Is it necessary to check it twice? It seems like a code duplication to me.

Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentController.php]: Ok, so how do we check if we can return 404 based on validation with the hasEntityAccess at line 493?

Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentController.php]: >  It should get 403, not 401.

I mean "...not 404" (notfound).

Comment (escopecz) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentController.php]: Here is an example how "not found" is implemented for landing pages:

https://github.com/mautic/mautic/blob/5.x/app/bundles/PageBundle/Controller/PageController.php#L496-L507

Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentController.php]: But this example does not check whether the user has permission to find out whether the entity exists or not.

Comment (escopecz) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentController.php]: No entity in Mautic checks whether the user has permission to find out whether the entity exists or not. I wouldn't over-complicated this :)

Comment (AlanWierzchonCA) [reply app/bundles/DynamicContentBundle/Controller/DynamicContentController.php]: Yes, secuirity issues are sometimes overcomplicated and the Department of Security gives many absurd points. 
As you wish, I have removed this access checking :)

**Review by escopecz — APPROVED** [review]
Amazing! The code is good to go and the tests are amazing! Thank you for taking care of this and be patient with me. 👍 


One last thing so this can be tested and merged. Can you please rebase it on top of 5.x branch and if you want to back-port it to Mautic 4 as well then create a duplicate PR against the 4.4 branch?

**Comment by AlanWierzchonCA** [comment]
Hi @escopecz, 

I made rebase these changes to 4.4
and created a duplicate PR against the 5.x #11480

**Pull request merged**